### PR TITLE
add auto-generated timestamp and sent_at (ruby-sdk)

### PR DESCRIPTION
**Pull request by bartes**
related https://github.com/castle/api/issues/1753
- replaced with_context with context_sanitizer class
- made hash options, params immutable
- added auto-generating timestamp when creating client from request
- added to_options method which automatically adds timestamp to options
- added possibility to either add timestamp to client options or pass it along with the query params

### CODE REVIEW

File: lib/castle/client.rb [hunk 1]
```diff
@@ -4,19 +4,28 @@ module Castle
...
+        new(
+          to_context(request, options),
+          to_options(options)
+        )
       end
 
-      def to_context(request, options = {})
-        default_context = Castle::DefaultContext.new(request, options[:cookies]).call
-        Castle::ContextMerger.new(default_context).call(options[:context] || {})
+      def to_context(request, context_options = {})
```
Comment (lluft) [inline lib/castle/client.rb]: why renaming to `context_options`, I think if any I would call it `additional_context` see other methods

File: lib/castle/client.rb [hunk 2]
```diff
@@ -4,19 +4,28 @@ module Castle
...
+        default_context = Castle::DefaultContext.new(request, context_options[:cookies]).call
+        Castle::ContextMerger.call(default_context, context_options[:context])
+      end
+
+      def to_options(options = {})
+        options[:timestamp] ||= Time.now.iso8601
```
Comment (lluft) [inline lib/castle/client.rb]: use `Time.now.utc.iso8601(3)`

File: lib/castle/client.rb [hunk 3]
```diff
@@ -25,6 +34,7 @@ def authenticate(options = {})
       options = Castle::Utils.deep_symbolize_keys(options || {})
 
       if tracked?
+        options[:timestamp] = @timestamp if @timestamp
```
Comment (lluft) [inline lib/castle/client.rb]: not sure if we should allow it for `authenticate` thoughts @wallin?

File: lib/castle/client.rb [hunk 4]
```diff
@@ -40,6 +50,7 @@ def identify(options = {})
       options = Castle::Utils.deep_symbolize_keys(options || {})
 
       return unless tracked?
+      options[:timestamp] = @timestamp if @timestamp
```
Comment (lluft) [inline lib/castle/client.rb]: I'd change it for all methods to `options[:timestamp] ||= ` that way you can also pass in a timestamp to identify/track

File: lib/castle/commands/authenticate.rb [hunk 5]
```diff
@@ -3,13 +3,18 @@
...
 
       def build(options = {})
         validate!(options)
-        build_context!(options)
+        context = ContextMerger.call(@context, options[:context])
+        context = ContextSanitizer.call(context)
 
-        Castle::Command.new('authenticate', options, :post)
+        Castle::Command.new('authenticate',
+                            options.merge(context: context, sent_at: Time.now.iso8601),
```
Comment (lluft) [inline lib/castle/commands/authenticate.rb]: `Time.now.utc.is8601(3)` also shouldn't this be the responsibility of `Castle::Command` will DRY it up and add responsibility to the Object making the request

File: lib/castle/commands/identify.rb [hunk 6]
```diff
@@ -3,13 +3,18 @@
...
 
       def build(options = {})
         validate!(options)
-        build_context!(options)
+        context = ContextMerger.call(@context, options[:context])
+        context = ContextSanitizer.call(context)
 
-        Castle::Command.new('identify', options, :post)
+        Castle::Command.new('identify',
+                            options.merge(context: context, sent_at: Time.now.iso8601),
```
Comment (lluft) [inline lib/castle/commands/identify.rb]: see comment on `authenticate`

File: lib/castle/commands/track.rb [hunk 7]
```diff
@@ -3,13 +3,18 @@
...
 
       def build(options = {})
         validate!(options)
-        build_context!(options)
+        context = ContextMerger.call(@context, options[:context])
+        context = ContextSanitizer.call(context)
 
-        Castle::Command.new('track', options, :post)
+        Castle::Command.new('track',
+                            options.merge(context: context, sent_at: Time.now.iso8601),
```
Comment (lluft) [inline lib/castle/commands/track.rb]: see comment on `authenticate`

**Comment by bartes** [comment]
@lluft thanks - corrected

**Comment by bartes** [comment]
as @wallin mentioned we don't need send it for `identify` (but allowing to send all of that to identify as well for other future calculations)

**Comment by lluft** [comment]
yeah Segment is sending it as well for identify

**Pull request merged**